saltnado: don't use gather_job_timeout as a timeout signal in job_not_running

oeuftete opened the pull request:
(Summary copied from #49567.  This was intended for Neon, but appears to have not made it off of the old `neon` branch?)

because the future was generated from tornado.gen.sleep, its eventual timer
still fires regardless of if the future is manually set with set_result. When
set_result is called twice (once manually by saltnado, once by tornado call_later),
a stacktrace like so is thrown:

```
[ERROR   ] Uncaught exception POST / (::1)
Traceback (most recent call last):
  File "/home/mphillips81/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tornado/web.py", line 1369, in _stack_context_handle_exception
    raise_exc_info((type, value, traceback))
  File "<string>", line 3, in raise_exc_info
  File "/home/mphillips81/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tornado/stack_context.py", line 314, in wrapped
    ret = fn(*args, **kwargs)
  File "/home/mphillips81/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tornado/gen.py", line 771, in later
    f.set_result(None)
  File "/home/mphillips81/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tornado/concurrent.py", line 254, in set_result
    self._set_done()
  File "/home/mphillips81/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tornado/concurrent.py", line 298, in _set_done
    for cb in self._callbacks:
TypeError: 'NoneType' object is not iterable

```
This patch attempts to fix that. @jacksontj would appreciate a review I'm not hugely familiar with the saltnado/tornado code here.

### What does this PR do?
fixes a stack thrown after every call.

fixes https://github.com/saltstack/salt/issues/49572

### Tests written?
No

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

mattp- committed:
saltnado: don't use gather_job_timeout as a timeout signal in job_not_running

because the future was generated from tornado.gen.sleep, its eventual timer
still fires regardless of if the future is manually set with set_result. When
set_result is called twice (once manually by saltnado, once by tornado call_later),
a stacktrace like so is thrown:

[ERROR   ] Uncaught exception POST / (::1)
Traceback (most recent call last):
  File "/home/mphillips81/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tornado/web.py", line 1369, in _stack_context_handle_exception
    raise_exc_info((type, value, traceback))
  File "<string>", line 3, in raise_exc_info
  File "/home/mphillips81/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tornado/stack_context.py", line 314, in wrapped
    ret = fn(*args, **kwargs)
  File "/home/mphillips81/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tornado/gen.py", line 771, in later
    f.set_result(None)
  File "/home/mphillips81/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tornado/concurrent.py", line 254, in set_result
    self._set_done()
  File "/home/mphillips81/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tornado/concurrent.py", line 298, in _set_done
    for cb in self._callbacks:
TypeError: 'NoneType' object is not iterable

This patch attempts to fix that.

Ch3LL commented:
appreciate you porting this to master. I know this was not your original PR but any chance you can write a regression test here? If not i can schedule some time to do so

Commit:
Reinstate some skipped tests

These seem to already pass on master, before the changes added in this
PR, so something over the last several years fixed whatever was broken.

Commit:
Save local client SaltAPIHandler::job_not_running

It's not clear to me why the local client will no longer be in the object the second time around.  This makes three failing tests pass.

Commit:
Address teardown warning in test_rest_tornado

Commit:
Add test for rest_tornado exception

oeuftete commented:
Regression test written.  I don't like that it has a 1.1-second sleep, but it's the best I can do right now.  Open to suggestions for improvement.

Also:
* restored tests from old PRs/issues #23623 and #34753, which seem to have resolved over the intervening years for whatever reason(s)
* fixed a test warning

oeuftete commented on tests/integration/netapi/rest_tornado/test_app.py:
```diff
@@ -705,9 +730,7 @@ def verify_event(future):
-            self.assertEqual(
-                event["data"]["post"], {"foo": salt.utils.stringutils.to_bytes("bar")}
-            )
+            self.assertEqual(event["data"]["post"], {"foo": "bar"})
```
The change here makes this test able to pass, but it still seems flaky in Jenkins, so I've left it skipping.  When it fails, it's with:

```text
Traceback (most recent call last):
  File "/tmp/kitchen/testing/tests/integration/netapi/rest_tornado/test_app.py", line 723, in test_post
    event = future.result()
  File "/tmp/kitchen/testing/salt/ext/tornado/concurrent.py", line 242, in result
    self._check_done()
Exception: DummyFuture does not support blocking for results

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/kitchen/testing/salt/ext/tornado/testing.py", line 137, in __call__
    result = self.orig_method(*args, **kwargs)
  File "/tmp/kitchen/testing/tests/integration/netapi/rest_tornado/test_app.py", line 727, in test_post
    resolve_future_timeout, exc
AssertionError: Failed to resolve future under 60 secs: DummyFuture does not support blocking for result
```

bryceml commented:
re-run all full

Ch3LL commented:
re-run all full

Ch3LL commented:
ping @dwoz can you review here since you have a lot of experience with the tornado code.

oeuftete commented:
All checks are passing on the latest rebase, but I am concerned that all of the tests in `tests/integration/netapi/rest_tornado` are being skipped now in Jenkins builds (and using `runtests.py` locally), slow or fast.

oeuftete commented:
I have run the `rest_tornado` tests manually, working around the issue in https://github.com/saltstack/salt/issues/57748 that does not select them in CI.  They all pass for me.

Commit:
Run pre-commit hooks on changes

oeuftete commented:
Not sure if the integration tests for this are running in CI yet (some slow, as well as https://github.com/saltstack/salt/issues/57748), but they seem OK manually.

```text
$ nox -e 'pytest-zeromq-3(coverage=False)' -- --run-slow -v tests/integration/netapi/rest_tornado
nox > Running session pytest-zeromq-3(coverage=False)
nox > Re-using existing virtualenv at .nox/pytest-zeromq-3-coverage-false.
nox > Session pytest-zeromq-3(coverage=False) was successful.
nox > Running session pytest-parametrized-3(coverage=False, crypto=None, transport='zeromq')
nox > Re-using existing virtualenv at .nox/pytest-parametrized-3-coverage-false-crypto-none-transport-zeromq.
nox > pip install --progress-bar=off -U setuptools<50.0.0
nox > Not overwritting already existing .nox/pytest-parametrized-3-coverage-false-crypto-none-transport-zeromq/lib/python3.6/site-packages/apt_inst.cpython-36m-x86_64-linux-gnu.so with /usr/lib/python3/dist-packages/apt_inst.cpython-36m-x86_64-linux-gnu.so
nox > Not overwritting already existing .nox/pytest-parametrized-3-coverage-false-crypto-none-transport-zeromq/lib/python3.6/site-packages/apt_pkg.pyi with /usr/lib/python3/dist-packages/apt_pkg.pyi
nox > Not overwritting already existing .nox/pytest-parametrized-3-coverage-false-crypto-none-transport-zeromq/lib/python3.6/site-packages/apt_inst.pyi with /usr/lib/python3/dist-packages/apt_inst.pyi
nox > Not overwritting already existing .nox/pytest-parametrized-3-coverage-false-crypto-none-transport-zeromq/lib/python3.6/site-packages/apt with /usr/lib/python3/dist-packages/apt
nox > Not overwritting already existing .nox/pytest-parametrized-3-coverage-false-crypto-none-transport-zeromq/lib/python3.6/site-packages/aptsources with /usr/lib/python3/dist-packages/aptsources
nox > Not overwritting already existing .nox/pytest-parametrized-3-coverage-false-crypto-none-transport-zeromq/lib/python3.6/site-packages/python_apt-1.6.5_ubuntu0.3.egg-info with /usr/lib/python3/dist-packages/python_apt-1.6.5_ubuntu0.3.egg-info
nox > Not overwritting already existing .nox/pytest-parametrized-3-coverage-false-crypto-none-transport-zeromq/lib/python3.6/site-packages/apt_offline_core with /usr/lib/python3/dist-packages/apt_offline_core
nox > Not overwritting already existing .nox/pytest-parametrized-3-coverage-false-crypto-none-transport-zeromq/lib/python3.6/site-packages/apt_pkg.cpython-36m-x86_64-linux-gnu.so with /usr/lib/python3/dist-packages/apt_pkg.cpython-36m-x86_64-linux-gnu.so
nox > Not overwritting already existing .nox/pytest-parametrized-3-coverage-false-crypto-none-transport-zeromq/lib/python3.6/site-packages/aptdaemon with /usr/lib/python3/dist-packages/aptdaemon
nox > pip install --progress-bar=off -r requirements/static/py3.6/linux.txt
nox > pip uninstall -y pytest-salt
nox > python -m pytest --rootdir /home/ken/working/git/github/saltstack/salt-two --log-file=artifacts/logs/runtests-20200918161150.222438.log --log-file-level=debug --show-capture=no -ra -s --transport=zeromq --run-slow -v tests/integration/netapi/rest_tornado
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.6.9, pytest-6.0.2, py-1.9.0, pluggy-0.13.0 -- /home/ken/working/git/github/saltstack/salt-two/.nox/pytest-parametrized-3-coverage-false-crypto-none-transport-zeromq/bin/python
cachedir: .pytest_cache
tempdir: /tmp/salt-tests-tmpdir
max open files; soft: 3072; hard: 1048576
rootdir: /home/ken/working/git/github/saltstack/salt-two, configfile: pytest.ini
plugins: salt-factories-0.92.0, helpers-namespace-2019.1.8, tempdir-2019.10.12
collected 23 items                                                                                                                                                         

tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_multi_local_async_post PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_multi_local_async_post_multitoken PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_post_no_auth PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_regression_49572 PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_root PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_simple_local_async_post PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_simple_local_async_post_no_tgt No minions matched the target. No command was sent, no jid was assigned.
PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_simple_local_post PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_simple_local_post_invalid_request PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_simple_local_post_no_tgt No minions matched the target. No command was sent, no jid was assigned.
PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_simple_local_post_only_dictionary_request PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_simple_local_post_only_dictionary_request_with_order_masters SKIPPED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_simple_local_runner_async_post PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestSaltAPIHandler::test_simple_local_runner_post PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestMinionSaltAPIHandler::test_get PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestMinionSaltAPIHandler::test_get_no_mid PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestMinionSaltAPIHandler::test_post PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestMinionSaltAPIHandler::test_post_with_client PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestMinionSaltAPIHandler::test_post_with_incorrect_client PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestJobsSaltAPIHandler::test_get PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestRunSaltAPIHandler::test_get PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestEventsSaltAPIHandler::test_get PASSED
tests/integration/netapi/rest_tornado/test_app.py::TestWebhookSaltAPIHandler::test_post SKIPPED

========================================================================= short test summary info ==========================================================================
SKIPPED [1] tests/integration/netapi/rest_tornado/test_app.py:349: Undetermined race condition in test. Temporarily disabled.
SKIPPED [1] tests/integration/netapi/rest_tornado/test_app.py:684: Skipping until we can devote more resources to debugging this test.
====================================================================== 21 passed, 2 skipped in 37.25s ======================================================================
nox > Session pytest-parametrized-3(coverage=False, crypto=None, transport='zeromq') was successful.
nox > Ran multiple sessions:
nox > * pytest-zeromq-3(coverage=False): success
nox > * pytest-parametrized-3(coverage=False, crypto=None, transport='zeromq'): success
```

Ch3LL commented:
re-run all full
